ext/POSIX/POSIX.xs

```diff
@@ -3345,30 +3345,51 @@ write(fd, buffer, nbytes)
 void
 abort()
 
-#ifdef I_WCHAR
-#  include <wchar.h>
-#endif
-
 int
-mblen(s, n)
-	char *		s
+mblen(s, n = ~0)
+	SV *		s
 	size_t		n
-    PREINIT:
-#if defined(USE_ITHREADS) && defined(HAS_MBRLEN)
-        mbstate_t ps;
-#endif
     CODE:
-#if defined(USE_ITHREADS) && defined(HAS_MBRLEN)
-        memset(&ps, 0, sizeof(ps)); /* Initialize state */
-        RETVAL = mbrlen(s, n, &ps); /* Prefer reentrant version */
-#else
-        /* This might prevent some races, but locales can be switched out
-         * without locking, so this isn't a cure all */
-        LOCALE_LOCK;
-
-        RETVAL = mblen(s, n);
-        LOCALE_UNLOCK;
+        errno = 0;
+        if (! SvOK(s)) {   /* Reset state */
+            /* The return value of mblen() when the first parameter is NULL
+             * indicates if the locale is stateless or stateful.  We otherwise
+             * use the more thread safe function mbrlen(), but that function
+             * doesn't give this information.  Thus we have to call mblen()
+             * itself, and we use a semaphore to minizimize races.  (The lock
+             * isn't fool-proof as the locale can be changed without a lock.)
+             * And we initialize the mbrlen() state separately.  We use
+             * mbrlen even in unthreaded perls so that the errno setting is
+             * uniform regardless of the configuration */
+            LOCALE_LOCK;
+            RETVAL = mblen(NULL, 0);
```

Comment:
not sure to read it correctly, but cannot it trigger an infinite loop there? by calling itself with `NULL` which will be false when checking `SvOK(s)`?

Reply:
There are two different  functions that appear to be named mblen(), but one actually has a different expanded name..  The call on line 3365 calls the libc version.  Code calling POSIX::mblen() gets the function in POSIX.xs.